onboard/catkin_ws/src/controls/src/drc_pid.cpp

```diff
@@ -0,0 +1,139 @@
+#include "drc_pid.h"
+
+#include <ros/ros.h>
+
+#include <memory>
+
+#include "controls_types.h"
+#include "controls_utils.h"
+
+PID::PID(){};
+
+PID::PID(const double &control_effort_min, const double &control_effort_max,
+         const PIDDerivativeTypesEnum &derivative_type, const double &error_ramp_rate, const PIDGainsMap &pid_gains) {
+    // Validate inputs
+    ROS_ASSERT_MSG(control_effort_min <= control_effort_max,
+                   "PID initialization error: Control effort min must be less than or equal to control effort max.");
+    ROS_ASSERT_MSG(ControlsUtils::pid_gains_map_valid(pid_gains),
+                   "PID initialization error: PID gains map is invalid.");
+    ROS_ASSERT_MSG(error_ramp_rate >= 0, "PID initialization error: Error ramp rate must be non-negative.");
+    ROS_ASSERT_MSG(ControlsUtils::value_in_pid_derivative_types_enum(derivative_type),
+                   "PID initialization error: Derivative type is invalid.");
+
+    // Set parameters
+    this->control_effort_min = control_effort_min;
+    this->control_effort_max = control_effort_max;
+    this->derivative_type = derivative_type;
+    this->error_ramp_rate = error_ramp_rate;
+    this->pid_gains = pid_gains;
+}
+
+void PID::set_pid_gain(const PIDGainTypesEnum &pid_gain_type, const double &value) {
+    // Set gains
+    this->pid_gains[pid_gain_type] = value;
+}
+
+const PIDGainsMap &PID::get_pid_gains() const { return pid_gains; }
+
+void PID::reset() {
+    // Reset integral to zero
+    integral = 0.0;
+}
+
+double PID::second_order_butterworth(const std::array<double, 3> &values,
+                                     const std::array<double, 3> &filtered_values) {
+    // Reference: Julius O. Smith III, Intro. to Digital Filters With Audio Applications.
+    // See https://ccrma.stanford.edu/~jos/filters/Example_Second_Order_Butterworth_Lowpass.html for a description of
+    // the math. Implementation adapted from https://bitbucket.org/AndyZe/pid/src/master/src/pid.cpp#lines-271.
+    return (1 / (1 + (c * c) + (1.414 * c))) *
+           (values.at(2) + (2 * values.at(1)) + values.at(0) - ((c * c) - (1.414 * c) + 1) * filtered_values.at(2) -
+            ((-2 * c * c) + 2) * filtered_values.at(1));
+}
+
+double PID::run(double error, double delta_time, PIDInfo &info, double provided_derivative) {
+    // If there are validation errors, an exception is not thrown to maintain continuous operation of controls in the
+    // event of a temporary error. Instead, the function returns 0 and prints an error message.
+
+    // If pid_gains is not valid, return 0 and print error
```

Comment:
If `pid_gains` can only be updated through `set_pid_gains`, why don't we validate `pid_gains` there instead of doing that every PID loop iteration?

Reply:
Validating `pid_gains` is necessary in `run` if the user initializes the class with the default constructor, in which `pid_gains` is an empty map and the `run` function, if called, would fail. If the user initializes the class with the other constuctor or has set all pid gains though `set_pid_gain`, then the `run` function would succeed.

However, what you are implying is correct – it is inefficient to check the validity of `pid_gains` in every iteration of `run`, especially since it is being run hundreds of times a second. Therefore, to avoid the edge case described above, I initialized all `pid_gains` to zero in the default constructor. Now, all pid gains should exist in `pid_gains` so the `run` function does not have to validate every time.